web/backend/routes/root.ts

```diff
@@ -202,12 +200,31 @@ export default async function (fastify: FastifyInstance) {
     return api.getRuntimeServices(request.params.pid)
   })
 
-  typedFastify.get('/runtimes/:pid/logs', {
-    schema: {
-      params: { type: 'object', properties: { pid: { type: 'number' } }, required: ['pid'] }
+  typedFastify.get<{ Params: PidParam }>('/runtimes/:pid/logs/ws', {
+    schema: { params: pidParamSchema, hide: true },
+    websocket: true
+  }, async (socket, { params: { pid } }) => {
+    try {
+      const clientStream = api.getRuntimeLiveLogsStream(pid)
+
+      socket.on('close', () => {
+        clientStream.destroy()
+      })
+
+      await pipeline(
+        clientStream,
+        split2(),
+        async function * (source) {
+          for await (const line of source) {
+            socket.send(line)
```

Comment:
we should try to avoid overloading the server.

You should wrap socket.send in a promise and use its callback: https://github.com/websockets/ws/blob/HEAD/doc/ws.md#websocketsenddata-options-callback.

Reply:
done d2f6a0b